Adjust slider ends to be more lenient during very fast sliders

peppy opened the pull request:
- [x] Depends on https://github.com/ppy/osu/pull/24965
- [x] ~~Update conversion tests with new output~~ Locally fix conversion tests for now
- Closes https://github.com/ppy/osu/issues/22805

I've removed all usage of the legacy tick in the osu! ruleset. This makes things a lot more sane in my eyes:

- The tail is mapped to the tail now
- The judgement logic in `DrawableSlider` reads better

If we go in this direction, the stable `expected-conversion` code will need to be updated, as it no longer matches. This can be seen from the test failures on this PR currently.

If this is seen as an overbearing or otherwise unwanted change, then the last two commits (62bcb6284232833fbec4f30a8f79bfa635771ddf and dd6d09189e21d7a58df891823ba1daf9e8625c83) can be reverted.

Another direction I explored was removing the code from `SliderEventGenerator` completely – the only thing relying on it as of this PR is osu!catch conversion. You can see how this would look on [this branch](https://github.com/ppy/osu/compare/master...peppy:osu:legacy-tick-go-boom?expand=1). I stopped exploring this direction as it feels like this would be less legible (having hidden logic in `JuiceStream`) but others may have different opinions.

<details>
<summary>Diff for better visualisation as used for videos below</summary>

```diff
diff --git a/osu.Game.Rulesets.Osu/Objects/Drawables/DrawableSliderTail.cs b/osu.Game.Rulesets.Osu/Objects/Drawables/DrawableSliderTail.cs
index 7daab820dd..e81ee84916 100644
--- a/osu.Game.Rulesets.Osu/Objects/Drawables/DrawableSliderTail.cs
+++ b/osu.Game.Rulesets.Osu/Objects/Drawables/DrawableSliderTail.cs
@@ -8,11 +8,13 @@
 using osu.Framework.Allocation;
 using osu.Framework.Graphics;
 using osu.Framework.Graphics.Containers;
+using osu.Framework.Graphics.Shapes;
+using osu.Game.Graphics.Sprites;
 using osu.Game.Rulesets.Objects;
 using osu.Game.Rulesets.Objects.Drawables;
-using osu.Game.Rulesets.Objects.Types;
 using osu.Game.Skinning;
 using osuTK;
+using osuTK.Graphics;
 
 namespace osu.Game.Rulesets.Osu.Objects.Drawables
 {
@@ -42,6 +44,8 @@ public partial class DrawableSliderTail : DrawableOsuHitObject, IRequireTracking
 
         private Container scaleContainer;
 
+        private Drawable box;
+
         public DrawableSliderTail()
             : base(null)
         {
@@ -68,7 +72,27 @@ private void load()
                     Children = new Drawable[]
                     {
                         // no default for this; only visible in legacy skins.
-                        CirclePiece = new SkinnableDrawable(new OsuSkinComponentLookup(OsuSkinComponents.SliderTailHitCircle), _ => Empty())
+                        CirclePiece = new SkinnableDrawable(new OsuSkinComponentLookup(OsuSkinComponents.SliderTailHitCircle), _ => box = new Container
+                        {
+                            Size = new Vector2(32),
+                            Children = new Drawable[]
+                            {
+                                new Box
+                                {
+                                    Size = new Vector2(32),
+                                    Anchor = Anchor.Centre,
+                                    Origin = Anchor.Centre,
+                                },
+                                new OsuSpriteText
+                                {
+                                    Text = "LegacyLastTick",
+                                    Anchor = Anchor.BottomCentre,
+                                    Origin = Anchor.TopCentre,
+                                    Scale = new Vector2(2),
+                                    Y = 5,
+                                }
+                            }
+                        })
                     }
                 },
             });
@@ -132,9 +156,15 @@ protected override void CheckForResult(bool userTriggered, double timeOffset)
             // The player needs to have engaged in tracking at any point after the tail leniency cutoff.
             // An actual tick miss should only occur if reaching the tick itself.
             if (timeOffset >= SliderEventGenerator.TAIL_LENIENCY && Tracking)
+            {
+                box.Colour = Color4.LimeGreen;
                 ApplyResult(r => r.Type = r.Judgement.MaxResult);
+            }
             else if (timeOffset >= 0)
+            {
+                box.Colour = Color4.Red;
                 ApplyResult(r => r.Type = r.Judgement.MinResult);
+            }
         }
 
         protected override void OnApply()
@@ -142,7 +172,7 @@ protected override void OnApply()
             base.OnApply();
 
             if (Slider != null)
-                Position = Slider.CurvePositionAt(HitObject.RepeatIndex % 2 == 0 ? 1 : 0);
+                Position = HitObject.Position - DrawableSlider.HitObject.Position;
         }
     }
 }
diff --git a/osu.Game.Rulesets.Osu/Objects/Slider.cs b/osu.Game.Rulesets.Osu/Objects/Slider.cs
index c62659d67a..2126ddf6c3 100644
--- a/osu.Game.Rulesets.Osu/Objects/Slider.cs
+++ b/osu.Game.Rulesets.Osu/Objects/Slider.cs
@@ -209,7 +209,7 @@ protected override void CreateNestedHitObjects(CancellationToken cancellationTok
                         {
                             RepeatIndex = e.SpanIndex,
                             StartTime = e.Time,
-                            Position = EndPosition,
+                            Position = Position + Path.PositionAt(e.PathProgress),
                             StackHeight = StackHeight
                         });
                         break;
diff --git a/osu.Game/Rulesets/Objects/SliderEventGenerator.cs b/osu.Game/Rulesets/Objects/SliderEventGenerator.cs
index 42c422a637..0b6f258565 100644
--- a/osu.Game/Rulesets/Objects/SliderEventGenerator.cs
+++ b/osu.Game/Rulesets/Objects/SliderEventGenerator.cs
@@ -108,7 +108,7 @@ public static class SliderEventGenerator
                 SpanIndex = finalSpanIndex,
                 SpanStartTime = startTime + (spanCount - 1) * spanDuration,
                 Time = startTime + totalDuration,
-                PathProgress = spanCount % 2,
+                PathProgress = finalProgress,
             };
         }
 

```

</details>

---

On very fast sliders, you now only need to be tracking somewhere in the last 36ms, rather than precisely at the (hidden) position of the last tick.

## Miss

https://github.com/ppy/osu/assets/191335/420acf91-d762-4718-b723-5f2687362a55

## Valid hits

https://github.com/ppy/osu/assets/191335/ea3e93b3-e4c1-41c8-be6d-d58dd4be64de

https://github.com/ppy/osu/assets/191335/4dc684ce-1dab-4133-ad7f-794c8d4a73cc

https://github.com/ppy/osu/assets/191335/d5d3b584-2446-4bca-ae86-766a1667a91c

peppy committed:
Apply NRT to `TestSceneSliderInput`

peppy committed:
Add gameplay test coverage around last tick of slider

This includes proposed changes as per
https://github.com/ppy/osu/issues/22805#issuecomment-1740377493.

peppy committed:
Adjust slider judgement logic to allow tracking anywhere after last tick

peppy committed:
Remove usage of `LastTick` in osu! ruleset

peppy committed:
Document remaining pieces of `LastTick` and add back legacy prefix for generation flow

peppy committed:
Add test coverage of tracking inside lenience period but not at tick or end

peppy committed:
Refactor tests to be easier to visually understand

bdach commented on osu.Game.Rulesets.Osu/Objects/Drawables/DrawableSliderTail.cs:
```diff
@@ -125,8 +126,15 @@ protected override void UpdateHitStateTransforms(ArmedState state)
+            // An actual tick miss should only occur if reaching the tick itself.
+            if (timeOffset >= SliderEventGenerator.TAIL_LENIENCY && Tracking)
+                ApplyResult(r => r.Type = r.Judgement.MaxResult);
+            else if (timeOffset >= 0)
```
Would probably be more comfortable with

```suggestion
            else if (timeOffset > 0)
```

Not sure it's gonna matter in practice; it might on legacy replays where all time instants are non-fractional.

bdach commented on osu.Game.Rulesets.Osu.Tests/TestSceneSliderInput.cs:
```diff
@@ -26,14 +25,53 @@ namespace osu.Game.Rulesets.Osu.Tests
+
+        private readonly List<JudgementResult> judgementResults = new List<JudgementResult>();
+
+        [TestCase(300, false)]
+        [TestCase(200, true)]
+        [TestCase(150, true)]
+        [TestCase(120, true)]
+        [TestCase(60, true)]
+        [TestCase(10, true)]
```
Would probably add a few more test cases just for fuller coverage:

```diff
diff --git a/osu.Game.Rulesets.Osu.Tests/TestSceneSliderInput.cs b/osu.Game.Rulesets.Osu.Tests/TestSceneSliderInput.cs
index 5ee72757d5..498399fd84 100644
--- a/osu.Game.Rulesets.Osu.Tests/TestSceneSliderInput.cs
+++ b/osu.Game.Rulesets.Osu.Tests/TestSceneSliderInput.cs
@@ -44,6 +44,8 @@ public partial class TestSceneSliderInput : RateAdjustedBeatmapTestScene
         [TestCase(120, true)]
         [TestCase(60, true)]
         [TestCase(10, true)]
+        [TestCase(0, true)]
+        [TestCase(-30, false)]
         public void TestTailLeniency(float finalPosition, bool hit)
         {
             performTest(new List<ReplayFrame>

```

peppy replied:
Added one, added the other commented out.

bdach commented:
> I've removed all usage of the legacy tick in the osu! ruleset. This makes things a lot more sane in my eyes

Generally tend to agree on this part for what it's worth.

bdach commented on osu.Game.Rulesets.Osu/Objects/Drawables/DrawableSliderTail.cs:
```diff
@@ -125,8 +126,15 @@ protected override void UpdateHitStateTransforms(ArmedState state)
+
+            // The player needs to have engaged in tracking at any point after the tail leniency cutoff.
+            // An actual tick miss should only occur if reaching the tick itself.
+            if (timeOffset >= SliderEventGenerator.TAIL_LENIENCY && Tracking)
```
One thing I didn't realise earlier is that this isn't 100% correct. If a slider is shorter than 72ms, the tail leniency actually ends midway through the slider, rather than a constant 36ms back, as per:

```csharp
            double finalSpanEndTime = Math.Max(startTime + totalDuration / 2, (finalSpanStartTime + spanDuration) + TAIL_LENIENCY);
```

in `LegacyLastTick` generation.

peppy replied:
Yeah, I noticed this one but forgot to reapply it here.

peppy replied:
Honestly not sure we want to apply this? If we're changing the behaviour, it seems more correct to remove this completely..

bdach replied:
I suppose. Probably still need a check for sliders shorter than 36ms just to avoid any weirdness with slider tails getting judged before the actual slider starts.

bdach replied:
@peppy is that "shorter than 36ms" check happening here or are we leaving it be as irrelevant? Just wanna make sure this hasn't been forgotten.

bdach replied:
Actually I guess the "check all preceding nested objects judged" thing covers that since, among others, the slider head is also a nested, so the tail won't be judged before the head is...

Walavouchey commented:
slider end hitsounds play too early in this branch

https://github.com/ppy/osu/assets/36758269/a3953486-65b0-4cda-b83d-2979bdae4a20

first half of the video is this branch, second half is the latest release

Natelytle commented:
https://github.com/ppy/osu/assets/92956514/0d62ee8b-67fa-4972-adc8-8324ba19c37e

It looks like it's impossible to hit 300s on fast enough buzz sliders

peppy commented:
> 
> 
> https://github.com/ppy/osu/assets/92956514/0d62ee8b-67fa-4972-adc8-8324ba19c37e
> 
> 
> 
> It looks like it's impossible to hit 300s on fast enough buzz sliders

See https://github.com/ppy/osu/pull/24966#discussion_r1341932320 👍

peppy commented on osu.Game.Rulesets.Osu/Objects/Drawables/DrawableSlider.cs:
```diff
@@ -256,7 +256,7 @@ public override void OnKilled()
         protected override void CheckForResult(bool userTriggered, double timeOffset)
         {
-            if (userTriggered || Time.Current < HitObject.EndTime)
+            if (userTriggered || !TailCircle.Judged)
```
```suggestion
            if (userTriggered || Time.Current < HitObject.EndTime || !TailCircle.Judged)
```

Should fix hitsounds

peppy committed:
Fix hitsounds playing too early on fast sliders

peppy committed:
Add a couple of extra tests

peppy committed:
Add test coverage of very short sliders

peppy commented:
@Natelytle could you please attach the beatmap, and if possible confirm it works on `master`? Also can you test without classic mod?

peppy committed:
Improve test output for judgement checking

Natelytle commented:
Sure, I'll test first thing in the morning tomorrow

peppy committed:
Fix slider tracking state being set to `false` after slider end

peppy committed:
Add note about broken test

peppy commented:
I've pushed some tentative fixes and test coverage of short sliders, but the tests still pass without the change. So will wait to hear back on the breaking sliders before investigating further.

peppy committed:
Change comparison of `timeOffset` to greater-than (in line with others)

Co-authored-by: Bartłomiej Dach <dach.bartlomiej@gmail.com>

peppy commented:
> slider end hitsounds play too early in this branch

@Walavouchey should be resolved via https://github.com/ppy/osu/pull/24966/commits/07207ffc322108beb76a6ca9bee9566581eec150.

Natelytle commented:
[Here's the test map I was using, it's a .olz that I turned into a .zip](https://github.com/ppy/osu/files/12783700/Camellia.-.Exit.This.Earth_s.Atomosphere.rrtyui.zip)

The sliders work fine on the latest master, and give the max amount of score with and without classic mod.
![image](https://github.com/ppy/osu/assets/92956514/a4e348e3-cec8-415f-99ff-396309308001)
![image](https://github.com/ppy/osu/assets/92956514/eee6a0a3-9b66-4a23-b672-4b9d15174620)

On the latest commit of this PR, the sliders no longer give 100s, but they don't give the max amount of score. 
![image](https://github.com/ppy/osu/assets/92956514/7c927e73-97c2-4d0d-83c7-dd5a0f5e8143)
![image](https://github.com/ppy/osu/assets/92956514/96029784-3177-4c55-b0e3-7c2ca96b2991)

bdach commented:
> On the latest commit of this PR, the sliders no longer give 100s, but they don't give the max amount of score

This is a new standardised score / "score V2" quirk. Compare: https://github.com/ppy/osu/pull/23866, https://github.com/ppy/osu/pull/23917.

Let's take the last kickslider. When simulating autoplay before the actual map, judgements are awarded in the following order:

```
[runtime] 2023-10-02 16:03:43 [verbose]: osu.Game.Rulesets.Osu.Objects.SliderHeadCircle @ 46847 judged as Great (simulating = True)
[runtime] 2023-10-02 16:03:43 [verbose]: osu.Game.Rulesets.Osu.Objects.SliderRepeat @ 46869.05882294505 judged as LargeTickHit (simulating = True)
[runtime] 2023-10-02 16:03:43 [verbose]: osu.Game.Rulesets.Osu.Objects.SliderRepeat @ 46891.1176458901 judged as LargeTickHit (simulating = True)
[runtime] 2023-10-02 16:03:43 [verbose]: osu.Game.Rulesets.Osu.Objects.SliderRepeat @ 46913.176468835154 judged as LargeTickHit (simulating = True)
[runtime] 2023-10-02 16:03:43 [verbose]: osu.Game.Rulesets.Osu.Objects.SliderRepeat @ 46935.23529178021 judged as LargeTickHit (simulating = True)
[runtime] 2023-10-02 16:03:43 [verbose]: osu.Game.Rulesets.Osu.Objects.SliderRepeat @ 46957.294114725264 judged as LargeTickHit (simulating = True)
[runtime] 2023-10-02 16:03:43 [verbose]: osu.Game.Rulesets.Osu.Objects.SliderRepeat @ 46979.352937670315 judged as LargeTickHit (simulating = True)
[runtime] 2023-10-02 16:03:43 [verbose]: osu.Game.Rulesets.Osu.Objects.SliderRepeat @ 47001.41176061537 judged as LargeTickHit (simulating = True)
[runtime] 2023-10-02 16:03:43 [verbose]: osu.Game.Rulesets.Osu.Objects.SliderRepeat @ 47023.47058356042 judged as LargeTickHit (simulating = True)
[runtime] 2023-10-02 16:03:43 [verbose]: osu.Game.Rulesets.Osu.Objects.SliderRepeat @ 47045.52940650548 judged as LargeTickHit (simulating = True)
[runtime] 2023-10-02 16:03:43 [verbose]: osu.Game.Rulesets.Osu.Objects.SliderRepeat @ 47067.58822945053 judged as LargeTickHit (simulating = True)
[runtime] 2023-10-02 16:03:43 [verbose]: osu.Game.Rulesets.Osu.Objects.SliderRepeat @ 47089.64705239558 judged as LargeTickHit (simulating = True)
[runtime] 2023-10-02 16:03:43 [verbose]: osu.Game.Rulesets.Osu.Objects.SliderTailCircle @ 47111.70587534063 judged as SmallTickHit (simulating = True)
```

When the _actual_ autoplay replay is played back, this however happens:

```
[runtime] 2023-10-02 16:03:50 [verbose]: osu.Game.Rulesets.Osu.Objects.SliderHeadCircle @ 46847 judged as Great (simulating = False)
[runtime] 2023-10-02 16:03:50 [debug]: Pressed (RightButton) handled by DrawableHitCircle+HitReceptor.
[runtime] 2023-10-02 16:03:50 [verbose]: osu.Game.Rulesets.Osu.Objects.SliderRepeat @ 46869.05882294505 judged as LargeTickHit (simulating = False)
[runtime] 2023-10-02 16:03:50 [verbose]: osu.Game.Rulesets.Osu.Objects.SliderRepeat @ 46891.1176458901 judged as LargeTickHit (simulating = False)
[runtime] 2023-10-02 16:03:50 [verbose]: osu.Game.Rulesets.Osu.Objects.SliderRepeat @ 46913.176468835154 judged as LargeTickHit (simulating = False)
[runtime] 2023-10-02 16:03:50 [verbose]: osu.Game.Rulesets.Osu.Objects.SliderRepeat @ 46935.23529178021 judged as LargeTickHit (simulating = False)
[runtime] 2023-10-02 16:03:50 [verbose]: osu.Game.Rulesets.Osu.Objects.SliderRepeat @ 46957.294114725264 judged as LargeTickHit (simulating = False)
[runtime] 2023-10-02 16:03:50 [verbose]: osu.Game.Rulesets.Osu.Objects.SliderRepeat @ 46979.352937670315 judged as LargeTickHit (simulating = False)
[runtime] 2023-10-02 16:03:50 [verbose]: osu.Game.Rulesets.Osu.Objects.SliderRepeat @ 47001.41176061537 judged as LargeTickHit (simulating = False)
[runtime] 2023-10-02 16:03:50 [verbose]: osu.Game.Rulesets.Osu.Objects.SliderRepeat @ 47023.47058356042 judged as LargeTickHit (simulating = False)
[runtime] 2023-10-02 16:03:50 [verbose]: osu.Game.Rulesets.Osu.Objects.SliderRepeat @ 47045.52940650548 judged as LargeTickHit (simulating = False)
[runtime] 2023-10-02 16:03:50 [verbose]: osu.Game.Rulesets.Osu.Objects.SliderRepeat @ 47067.58822945053 judged as LargeTickHit (simulating = False)
[runtime] 2023-10-02 16:03:50 [verbose]: osu.Game.Rulesets.Osu.Objects.SliderTailCircle @ 47076.03247229452 judged as SmallTickHit (simulating = False)
[runtime] 2023-10-02 16:03:50 [verbose]: osu.Game.Rulesets.Osu.Objects.SliderRepeat @ 47089.64705239558 judged as LargeTickHit (simulating = False)
```

Large ticks affect combo, small ticks do not. Because the combo portion of "score V2" contains a sort of combo feedback (by incorporating `ComboAfterJudgement + 1` into the formula), judging the non-combo-affecting tail circle before the combo-affecting slider repeat _decreases the value of the slider repeat_, as it doesn't get the extra 1 combo it would if the tail circle was judged after the repeat.

Possible solutions:

- When simulating autoplay, reorder judgements such that tail circles are judged as early as possible (fixes the immediate issue, but may cause score to get..... weird - stuff like doing things to hit a slider tail _late_ etc)
- Do not allow the slider tail to be judged before the last repeat.

peppy commented:
> Possible solutions:
> 
> * When simulating autoplay, reorder judgements such that tail circles are judged as early as possible (fixes the immediate issue, but may cause score to get..... weird - stuff like doing things to hit a slider tail _late_ etc)
> * Do not allow the slider tail to be judged before the last repeat.

I think the second makes sense, to the point I naively assumed that was already the case (but I guess not).

Personally I'd have the assumption that nested hit objects are always judged sequentially. Wonder if it's too late to enforce this? Discussing with @smoogipoo I think we both agree on this one.

peppy committed:
Add extra test coverage to `TestSceneOsuModAutoplay` to cover fail case

Basically the slider needs to be slightly longer for this test to
correctly fail in headless tests, in conjunction with the new slider
tail leniency.

This is due to headless tests running at a fixed frame interval, and
these timings being *tight*.

peppy committed:
Fix potential out-of-order tail and ticks

peppy committed:
Fix one case of difficulty calculation no longer accounting for leniency

peppy committed:
Rename weird diffcalc parameter name

peppy committed:
Merge branch 'master' into legacy-tick-test-coverage

peppy committed:
Fix difficulty calculation not correct handling slider leniency anymore

smoogipoo commented:
Updated sheet after the recent commit: https://docs.google.com/spreadsheets/d/1MB3PHT-5o0bwciVX4RaUJfXQj_tThyXbFmFjnHUOVco/edit#gid=1053465849

The changes are still larger than I expected... 17% pp losses on one beatmap. It's worth checking to make sure of the cause here.

As discussed, the current theory is that these are sliders where there's a normal tick after the "last tick". Thinking about it a bit more, `master` may be correct here, and I question whether this PR is handling things correctly at all, because I'm fairly certain osu!stable doesn't discriminate between ticks - it only gets the next tick (by time) and only if it's the last tick (by time) in the list does it apply a special judgement (see https://github.com/peppy/osu-stable-reference/blob/master/osu!/GameplayElements/HitObjects/Osu/SliderOsu.cs#L2039-L2044 ).

peppy commented:
Keep in mind that this PR *changes the game behaviour* to be something completely different from stable, so I would expect changes. As in we're not going to end up with a zero-change spreadsheet.

I'll go through the largest changes and provide analysis on what is causing them, regardless.

peppy committed:
Merge branch 'master' into legacy-tick-test-coverage

peppy committed:
Add note about early return in `DrawableSliderBall.Update`

peppy committed:
Add better commenting around legacy last tick edge case

peppy commented:
Taking a moment to address the SR changes found in the spreadsheet on 5 beatmaps, I've looked into the [beatmap](https://osu.ppy.sh/beatmapsets/881996#osu/1844019) with the largest change in SR. We come across sliders like this:

00:57:383 (1,2,3,4,5) - 

![osu! 2023-10-04 at 05 15 11](https://github.com/ppy/osu/assets/191335/1371d31a-a2bc-43b6-8486-628529f16b36)

Taking the first of these sliders as an example, the length of the slider is 43 ms, which will trigger the "lesser lenience" as explained in the [updated comments](https://github.com/peppy/osu/commit/cfb18e18c060f4270925b544fa545518ac48dea6) with quite some detail. The lenience will thus be 21.5 ms, occurring halfway on the slider's path:

![osu! 2023-10-04 at 05 16 52](https://github.com/ppy/osu/assets/191335/185d1e5d-8730-45fd-8ff3-1b2c5ee13503)

If we adjust the follow circle to show where the "actual" follow region is by removing its animation, it looks like this:

![osu! 2023-10-04 at 05 19 30](https://github.com/ppy/osu/assets/191335/f88de469-bad5-4499-8fb0-da4361ea72e8)

With the new changes, the lenience would still be 36 ms, meaning the user only needs to hold until 7 ms into the slider. Visually:

![osu! 2023-10-04 at 05 22 28](https://github.com/ppy/osu/assets/191335/a3efd150-f63a-42f2-a2a3-df20ad129ab8)

Both cases should **not require positional mouse movement by the user**, so you may wonder why difficulty calculation (which is supposed to take minimum movement requirements into consideration) is different. That's where we check the diffcalc code and find it doesn't use the full `2.4f` tracking radius of the follow circle, but a reduced `1.8f`:

https://github.com/ppy/osu/blob/8d919912149a4853120612b67944492c27baa611/osu.Game.Rulesets.Osu/Difficulty/Preprocessing/OsuDifficultyHitObject.cs#L25

This is how things look with stable-style-lenience at `1.8f`. Note that the follow circle now no longer overlaps the starting circle and *does* imply user mouse movement requirements:

![osu! 2023-10-04 at 05 25 52](https://github.com/ppy/osu/assets/191335/57b10b64-0e60-49b1-9469-1b6438e0e50a)

So it's arguable whether this is something that needs to be resolved here. Here's some points:

- In the cited beatmap, users are unlikely to move their mouse (since they don't need to), so in actual fact the diffcalc on `master` using the `1.8f` scaled follow circle is likely less accurate
- If we are considering backporting the changes in this branch to stable, we'll want diffcalc to match the new behaviour.

I'll add isolated diffcalc test coverage with this specific slider for anyone who wants to play along at home.

peppy committed:
Add test coverage of very fast slider

peppy commented:
d14d885 adds test coverage of the case in https://github.com/ppy/osu/pull/24966#issuecomment-1746162961 cc/ @smoogipoo 

Currently the tests are set using the values from `master` for reference.

peppy committed:
Fix regression in `Tracking` state handling

It turns out multiple components depend on `Tracking` eventually
becoming `false` at the end of a slider. With my previous changes, this
was no longer the case (as could be seen by the legacy cursor particles
test failure, and heard by slider slide taking too long to stop).

apollo-dw commented:
In my eyes (not speaking for the rest of the pp committee), I'm fine with either proceeding without any changes to difficulty calculation, or either just removing the notion of "assumed slider radius". 

Leaning towards the former though, just because the magnitude of such a change will be way greater than the 5 maps affected here (especially when I don't see the changes here as extraordinarily major)

peppy commented:
> In my eyes (not speaking for the rest of the pp committee), I'm fine with either proceeding without any changes to difficulty calculation, or either just removing the notion of "assumed slider radius".

So basically not applying the new "correct" approach as per lazer logic unless we backport this to stable, and having it "incorrect" but also unchanged from existing expectations?

apollo-dw commented:
To clarify: I'm fine with implementing the new lazer leniency mechanic whilst also not changing difficulty calculation for now, regardless of whether or when it gets backported to stable. The reason why I'm not concerned about it is because I don't think these changes in values are either game-breaking or widespread, so I'm happy to deal with it later and just ignore it for now. In other words, low priority.

smoogipoo commented:
New spreadsheet https://docs.google.com/spreadsheets/d/1QBwyQL_rW4vidHoJvhcKeeD2ehh7n0VdzZLYubMR2nA/edit#gid=782412943

smoogipoo commented:
!diffcalc

peppy commented:
!diffcalc

peppy committed:
Amend diffcalc to use something closer to the original calculation for now

peppy commented:
The previous diff was wrong, because it turns out that in the previous code, the last normal tick could come *after* the legacyLastTick, resulting in the `[^1]` operation coming upon the later last tick. This is more correctly:

```diff
diff --git a/osu.Game.Rulesets.Osu/Difficulty/Preprocessing/OsuDifficultyHitObject.cs b/osu.Game.Rulesets.Osu/Difficulty/Preprocessing/OsuDifficultyHitObject.cs
index 5c2c85d69e..f37452a79c 100644
--- a/osu.Game.Rulesets.Osu/Difficulty/Preprocessing/OsuDifficultyHitObject.cs
+++ b/osu.Game.Rulesets.Osu/Difficulty/Preprocessing/OsuDifficultyHitObject.cs
@@ -215,13 +215,20 @@ private void computeSliderCursorPosition(Slider slider)
             if (slider.LazyEndPosition != null)
                 return;
 
-            double trackingEndTime = Math.Max(
-                // SliderTailCircle always occurs at the final end time of the slider, but the player only needs to hold until within a lenience before it.
-                slider.NestedHitObjects.OfType<SliderTailCircle>().Single().StartTime + SliderEventGenerator.TAIL_LENIENCY,
-                // There's an edge case where one or more ticks/repeats fall within that leniency range.
-                // In such a case, the player needs to track until the final tick or repeat.
-                slider.NestedHitObjects.LastOrDefault(n => n is not SliderTailCircle)?.StartTime ?? double.MinValue
-            );
+            // This commented version is actually correct by the new lazer implementation, but intentionally held back from
+            // difficulty calculator to preserve known behaviour.
+            // double trackingEndTime = Math.Max(
+            //     // SliderTailCircle always occurs at the final end time of the slider, but the player only needs to hold until within a lenience before it.
+            //     slider.Duration + SliderEventGenerator.TAIL_LENIENCY,
+            //     // There's an edge case where one or more ticks/repeats fall within that leniency range.
+            //     // In such a case, the player needs to track until the final tick or repeat.
+            //     slider.NestedHitObjects.LastOrDefault(n => n is not SliderTailCircle)?.StartTime ?? double.MinValue
+            // );
+
+            double trackingEndTime = Math.Max(Math.Max(
+                slider.StartTime + slider.Duration + SliderEventGenerator.TAIL_LENIENCY,
+                slider.StartTime + slider.Duration / 2
+            ), slider.NestedHitObjects.OfType<SliderTick>().Last().StartTime);
 
             slider.LazyTravelTime = trackingEndTime - slider.StartTime;
 

```

This reduces the gap, and I've confirmed that `trackingEndTime` matches 1:1 with `master`. The remaining difference is due to the `for` loop just below this code, which is no longer getting tossed around by the `LegacyLastTick` and `LastTick` being mis-ordered and causing backward movement.

Given the above, I'm happy where this PR is. All that remains is a decision over whether we apply the local fix (aka the diff in this comment) to diffcalc or not.

@smoogipoo I'll push the diff to this branch for now, can you run a diffcalc sheet run?

peppy committed:
Fix potential failure when slider has no ticks

peppy commented:
We're down to only three beatmaps with SR changes now! Of note, there's one new beatmap showing up now which wasn't there previously (`beatmap_id=2404722`). I'm not sure if this is still worth investigating.

Previous: https://docs.google.com/spreadsheets/d/1QBwyQL_rW4vidHoJvhcKeeD2ehh7n0VdzZLYubMR2nA/edit#gid=1375311656
Now: https://docs.google.com/spreadsheets/d/1AFaVpvIMtt7Te70KOb3xx1U11KSVNC6TabQwOAZqtB0/edit#gid=1193118755

There's still, on one or two beatmaps, around a 10% pp gain at most. pp losses are down from 10% to <1%. If we want to keep things roughly untouched in terms of pp calculation, I'd propose this is a good middle ground?

If we want to make things match perfectly, the only way forward is to insert and re-sort the nested objects before running the `for` loop in `OsuDifficultyHitObject` to get 1:1 parity. @apollo-dw @smoogipoo thoughts?

peppy committed:
Add reordering support to match existing diffcalc 100%

peppy commented:
I've made a final change which should bring 1:1 parity with diffcalc, if that's what we want. Any of the last three commits can be reverted depending on the decided direction.

I also attempted to fix the conversion mapping stuff but I dunno if it's even feasible to get right. Here's what I have, but there are some new test failures which are probably related to the same thing as in the diffcalc logic and require similar fixes but to what end?

```diff
diff --git a/osu.Game.Rulesets.Osu.Tests/OsuBeatmapConversionTest.cs b/osu.Game.Rulesets.Osu.Tests/OsuBeatmapConversionTest.cs
index 11e69dc133..d2cba2a686 100644
--- a/osu.Game.Rulesets.Osu.Tests/OsuBeatmapConversionTest.cs
+++ b/osu.Game.Rulesets.Osu.Tests/OsuBeatmapConversionTest.cs
@@ -44,13 +44,25 @@ protected override IEnumerable<ConvertValue> CreateConvertValue(HitObject hitObj
                     break;
             }
 
-            static ConvertValue createConvertValue(OsuHitObject obj) => new ConvertValue
+            static ConvertValue createConvertValue(OsuHitObject obj)
             {
-                StartTime = obj.StartTime,
-                EndTime = obj.GetEndTime(),
-                X = obj.StackedPosition.X,
-                Y = obj.StackedPosition.Y
-            };
+                double startTime = obj.StartTime;
+                double endTime = obj.GetEndTime();
+
+                if (obj is SliderTailCircle)
+                {
+                    startTime += SliderEventGenerator.TAIL_LENIENCY;
+                    endTime += SliderEventGenerator.TAIL_LENIENCY;
+                }
+
+                return new ConvertValue
+                {
+                    StartTime = startTime,
+                    EndTime = endTime,
+                    X = obj.StackedPosition.X,
+                    Y = obj.StackedPosition.Y
+                };
+            }
         }
 
         protected override Ruleset CreateRuleset() => new OsuRuleset();

```

bdach commented:
Two of the failures are fixed by

```diff
diff --git a/osu.Game.Rulesets.Osu.Tests/OsuBeatmapConversionTest.cs b/osu.Game.Rulesets.Osu.Tests/OsuBeatmapConversionTest.cs
index d2cba2a686..c5a0b175ec 100644
--- a/osu.Game.Rulesets.Osu.Tests/OsuBeatmapConversionTest.cs
+++ b/osu.Game.Rulesets.Osu.Tests/OsuBeatmapConversionTest.cs
@@ -3,6 +3,7 @@
 
 using System;
 using System.Collections.Generic;
+using System.Linq;
 using NUnit.Framework;
 using osu.Framework.Utils;
 using osu.Game.Rulesets.Objects;
@@ -33,8 +34,13 @@ protected override IEnumerable<ConvertValue> CreateConvertValue(HitObject hitObj
             switch (hitObject)
             {
                 case Slider slider:
+                    var objects = new List<ConvertValue>();
+
                     foreach (var nested in slider.NestedHitObjects)
-                        yield return createConvertValue((OsuHitObject)nested);
+                        objects.Add(createConvertValue((OsuHitObject)nested));
+
+                    foreach (var obj in objects.OrderBy(cv => cv.StartTime))
+                        yield return obj;
 
                     break;
 

```

I'm not sure what to do with the third one, though. It seems like in `old-stacking` the slider tail is... 6.2ms off time-wise from where it should be? I have no idea why that would be the case, and I'm not entirely sure it's worth looking into.

Maybe the thing to do here is just remove slider tails from the mapping entirely. Or at least ignore their time values.

peppy commented:
Nice one. Want to commit that diff?

bdach commented:
I mean as stated there'll be one failure left, so dunno if that'll do anything...

bdach committed:
Fix conversion test failures

bdach committed:
Add inline commentary about workarounds in beatmap conversion test

bdach commented:
I think I fixed the remaining test failures, let's see what CI thinks...

Other than that I think we can try getting this in and seeing what the feedback is. I for one don't personally have any reason to hold this back further.

smoogipoo reviewed (dismissed):
LGTM, waiting on diffcalc spreadsheet run: https://github.com/ppy/osu/actions/runs/6546185309

smoogipoo committed:
Merge branch 'master' into legacy-tick-test-coverage

peppy committed:
Comment flaky tests for now

Same issue as other commented test (clock precision not high enough).

Will probably want a solution to this at some point.

peppy commented:
Diff calc re-run: https://github.com/ppy/osu/actions/runs/6546185309

peppy commented:
These tests are fucked on CI. FWIW I can't repro the failures locally.

Shall we just disable them all for now? Is this going to be blocked further while I refactor framework to allow adjustable test precision or similar?

bdach commented:
I'm okay with disable or retry.

peppy commented:
Let's see if retry actually works.

peppy committed:
Attempt to fix flaky tests with egregious retry rule

peppy commented:
> Diff calc re-run: https://github.com/ppy/osu/actions/runs/6546185309

Looks like there's still three SR changes (lowest yet). I guess we're going to have to look into the remaining discrepancies? 😮‍💨

bdach commented:
Egregiou retry also didn't help either somehow... I'm not sure if that works with parameterised testing but at this point probably best to ignore.

peppy committed:
Ignore instead of retry tests

peppy committed:
Merge branch 'master' into legacy-tick-test-coverage

peppy commented:
The remaining issues are floating point precision loss on a `double` (around `0.00000000008` difference on 45 sliders in the case of beatmap_id `1398809`).

![CleanShot 2023-10-20 at 08 55 34](https://github.com/ppy/osu/assets/191335/0ee5c75b-6c83-4531-8da7-b868ea42fe45)

Note the immediate window. The new method uses the "more precise" version.

Here's a diff for the above beatmap: https://gist.github.com/peppy/eb58fe462980a6b543bd06ee0d549647/revisions
Created using this patch: 

```diff
diff --git a/osu.Game.Rulesets.Osu/Difficulty/Preprocessing/OsuDifficultyHitObject.cs b/osu.Game.Rulesets.Osu/Difficulty/Preprocessing/OsuDifficultyHitObject.cs
index 488d1e2e9f..e0526a4136 100644
--- a/osu.Game.Rulesets.Osu/Difficulty/Preprocessing/OsuDifficultyHitObject.cs
+++ b/osu.Game.Rulesets.Osu/Difficulty/Preprocessing/OsuDifficultyHitObject.cs
@@ -307,6 +307,8 @@ private void computeSliderCursorPosition(Slider slider)
                 if (i == nestedObjects.Count - 1)
                     slider.LazyEndPosition = currCursorPosition;
             }
+
+            Console.WriteLine($"{slider.StartTime} {slider.LazyTravelTime} {slider.LazyEndPosition}");
         }
 
         private Vector2 getEndCursorPosition(OsuHitObject hitObject)

```

I'd argue there's no point investigating further and this change is good to go. If there change in PP is seen as an issue, it should be fixed at the diffcalc end by making the algorithm less susceptible to floating point differences like this.

peppy commented:
From StanR:

> this seems fine-ish imo, if the change is important enough it's fine to break the diffcalc a little - it can get fixed later if the issue is bad or undesirable, as for these changes I think it's just fine, the maps in question are underweighted anyway
>
> I think you shouldn't bother with diffcalc pretty much at all and just ping pp committees when changes affecting diffcalc are being rolled out

So let's go ahead with this as is.
